### PR TITLE
storage: consolidate how kafka and postgres connect to ssh tunnels

### PR DESCRIPTION
**Pull request by guswynn**
(this was originally pulled out from https://github.com/MaterializeInc/materialize/pull/22718 and has some review resolved)

We lift the ssh tunnel manager out of storage-client crate into the ssh-util crate, and force users to provide the keypair, not just a secrets reader. We then use the manager for postgres; before, postgres was making a new ssh tunnel per-client. Now, we won't make additional ssh tunnels, which means we can manage resources more consistently.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by mjibson — APPROVED** [review]
Adapter parts lgtm.

**Review by benesch — APPROVED** [review]
This looks really solid. :sob-love: Thanks for iterating on this one, @guswynn!

To avoid another code review roundtrip I pushed up two commits:

  1. I removed the `mz-repr`/`GlobalId` dependency from the `mz-ssh-util` crate by keying SSH tunnels by their config, rather than their connection ID. This also makes the API nice and clean, as the caller isn't responsible for managing a connection ID with some undocumented uniqueness properties. This commit deserves a quick look before you merge.
  2. This one's just a nitpick: I reordered the `ssh_tunnel_manager` parameter to be consistently the first parameter in all functions where it appears. As a matter of style I try to keep "context" parameters (like parameters of type `&ConnectionContext`) as the first parameters, and `ssh_tunnel_manager` functions very similarly to `ConnectionContext`.
  3. I pinned the `sshd` image to v1.7.1, which claims to work on ARM.

File: misc/python/materialize/mzcompose/services/ssh_bastion_host.py [hunk 1]
```diff
@@ -28,10 +28,11 @@ def __init__(
             MZ_ROOT / "misc" / "images" / "sshd" / "setup.sh",
             loader.composition_path,
         )
+
         super().__init__(
             name=name,
             config={
-                "image": "panubo/sshd:1.5.0",
+                "image": "quay.io/panubo/sshd",
```
[the remark this answers is not in the document]
Comment (benesch) [reply misc/python/materialize/mzcompose/services/ssh_bastion_host.py]: Handled in commit 3.

[the remark this answers is not in the document]
Comment (guswynn) [reply misc/python/materialize/mzcompose/services/ssh_bastion_host.py]: thanks!

**Comment by guswynn** [comment]
Thanks @benesch ! merging today!

**Pull request merged**